ledger/redis/interest.go

```diff
@@ -142,3 +157,164 @@ func SetUserInterest(
 
 	return redis.TryTransaction(ctx, rdb, txFunc, backoff.NewExponentialBackOff(backoff.WithMaxElapsedTime(timeout)), watch...)
 }
+
+func setUserInterestTx(ctx context.Context, tx redis.Cmdable, userID string, interest *types.Interest) error {
+	key := UserInterestKey(userID)
+	values := map[string]interface{}{
+		InterestEarned:      interest.Earned,
+		InterestOwed:        interest.Owed,
+		InterestClaimed:     interest.Claimed,
+		InterestPaid:        interest.Paid,
+		InterestLastUpdated: interest.LastUpdated,
+	}
+	if _, err := tx.HMSet(ctx, key, values).Result(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func AccrueLendingInterest(ctx context.Context, rdb redis.Client, timeout time.Duration, userID string, assetData helpers.AssetData, flatRate float64) (*gtypes.Transaction, error) {
+	// this has to be calculated within one transaction, we don't want positions changing between reads etc.
+	// as this could lead to inconsistencies
+	watch := []string{
+		UserInterestKey(userID),
+		UserPositionKey(userID),
+		UserInterestKey(MODULE),
+		ModulePositionKey(),
+	}
+
+	var accrualTransaction *gtypes.Transaction
+
+	txFunc := func(tx *redisv9.Tx) error {
+		transaction, err := accrueLendingInterestTx(ctx, tx, userID, assetData, flatRate)
+		if err != nil {
+			return err
+		}
+
+		accrualTransaction = transaction
+		return nil
+	}
+
+	err := redis.TryTransaction(ctx, rdb, txFunc, backoff.NewExponentialBackOff(backoff.WithMaxElapsedTime(timeout)), watch...)
+	if err != nil {
+		return nil, err
+	}
+
+	return accrualTransaction, nil
+}
+
+func AccrueAllLendingInterest(ctx context.Context, rdb redis.Client, timeout time.Duration, assetData helpers.AssetData, flatRate float64, watch []string, users ...string) ([]*gtypes.Transaction, error) {
+	watch = append(watch, UserInterestKey(MODULE), ModulePositionKey())
+	var accrualTransactions []*gtypes.Transaction
+
+	txFunc := func(tx *redisv9.Tx) error {
+		for _, userID := range users {
+			transaction, err := accrueLendingInterestTx(ctx, tx, userID, assetData, flatRate)
+			if err != nil {
+				return err
+			}
+
+			accrualTransactions = append(accrualTransactions, transaction)
+		}
+
+		return nil
+	}
+
+	err := redis.TryTransaction(ctx, rdb, txFunc, backoff.NewExponentialBackOff(backoff.WithMaxElapsedTime(timeout)), watch...)
+	if err != nil {
+		return nil, err
+	}
+
+	return accrualTransactions, nil
+}
+
+func accrueLendingInterestTx(ctx context.Context, tx redis.Cmdable, userID string, assetData helpers.AssetData, flatRate float64) (*gtypes.Transaction, error) {
+	now := time.Now()
+
+	// first get the module positions
+	var (
+		moduleInterest     types.Interest
+		accrualTransaction *gtypes.Transaction
+	)
+	interest, err := getInterestTx(ctx, tx, MODULE)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(interest) > 0 {
+		moduleInterest = interest[0]
+	}
+
+	// then get the user's positions
+	userPositionsCmd, err := GetUsersPositionCmd(ctx, tx, userID)
```

Comment:
Don't we have a GetUserPositionTx?

Reply:
Not in utils, maybe one was written in another service, but in the shared library we just have this one.